### PR TITLE
Add LaunchPad connector

### PR DESCRIPTION
**Pull request by bwitherspoon**
The LaunchPad / BoosterPack connector is found on TI LaunchPad development kits and BoosterPack expansion modules. This PR adds support for the LaunchPad connector which allows shields (e.g. BoosterPacks) to use aliases for the connector interfaces and mappings for the GPIOs. This is similar to what is being done for the Arduino connectors.

### CODE REVIEW

**Review by dbkinder — CHANGES_REQUESTED** [review]
Should there be documentation as well?

**Comment by bwitherspoon** [comment]
> Should there be documentation as well?

I haven't included any documentation in this PR. Do I need to add some?

I originally had this based on another PR which has now been merged and this PR rebased. Because of this several CODEOWNERS got pulled in due to those commits.

**Review by dbkinder — APPROVED** [review]
if no doc changes are needed, I'll get out of the way...

File: dts/bindings/gpio/launchpad-header.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+description: >
+    This is a representation of GPIO pin nodes exposed as headers on LaunchPads
+
+properties:
+    compatible:
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "launchpad-header"
```
Comment (galak) [inline dts/bindings/gpio/launchpad-header.yaml]: probably best to call this "ti,launchpad-header".

File: dts/bindings/gpio/launchpad-header.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      generation: define
```
Comment (galak) [inline dts/bindings/gpio/launchpad-header.yaml]: remove "generation:"

File: dts/bindings/gpio/launchpad-header.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2019 Brett Witherspoon
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+---
+title: LaunchPad GPIO header
+version: 0.1
```
Comment (galak) [inline dts/bindings/gpio/launchpad-header.yaml]: remove version

File: dts/bindings/gpio/launchpad-header.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2019 Brett Witherspoon
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+---
```
Comment (galak) [inline dts/bindings/gpio/launchpad-header.yaml]: remove `---`

**Review by galak — CHANGES_REQUESTED** [review]
Looks generally fine, a few cleanups.  I assume there are some "shields" you plan on adding?

File: dts/bindings/gpio/launchpad-header.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "launchpad-header"
+      generation: define
+
+    gpio-map:
+      type: compound
+      category: required
+
```
Comment (vanti) [inline dts/bindings/gpio/launchpad-header.yaml]: I am a newbie when it comes to dts files. But looking at the arduino ones, should "#gpio-cells" be listed as a property here? The property is being used on line 10 of the dtsi files.

Comment (vanti) [reply dts/bindings/gpio/launchpad-header.yaml]: LGTM otherwise

Comment (galak) [reply dts/bindings/gpio/launchpad-header.yaml]: I introduced `gpio-nexus.yaml` so including that like `arduino-header-r3.yaml` does will cover it.

**Comment by galak** [comment]
@bwitherspoon plans to update this?

**Comment by bwitherspoon** [comment]
I had plans to add some shields, but this work has stalled. I am going to close this PR and reopen when I have time to complete the shields.

**Comment by bwitherspoon** [comment]
I have started submitting PRs for the various shields I have worked with. This GPIO mapping is essential for configuring the interrupt lines on the various boards.

I have updated this connector to imitate how the arduino connector is now done and in the process addressed the mentioned issues.

**Comment by bwitherspoon** [comment]
Also are there any opinions on if this should be called BoosterPack connector instead of LaunchPad connector?

**Comment by vanti** [comment]
> Also are there any opinions on if this should be called BoosterPack connector instead of LaunchPad connector?

I think I have a preference for BoosterPack connector given it is referred to as "BoosterPack plug-in module connectors" on the [page for the TI LaunchPad](http://www.ti.com/tool/LAUNCHXL-CC1352R1). I also see some existing documentation (e.g. boards/arm/cc26x2r1_launchxl/doc/index.rst) where it is referred to as BoosterPack connector, so it is good to be consistent.

File: dts/bindings/gpio/ti,launchpad-header.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2019 Brett Witherspoon
+# SPDX-License-Identifier: Apache-2.0
+
+title: TI LaunchPad GPIO header
+
+description: |
+    This is a representation of GPIO pin nodes exposed as headers on TI LaunchPads
+
```
Comment (erwango) [inline dts/bindings/gpio/ti,launchpad-header.yaml]: Can you provide more documentation  (cf recent https://github.com/zephyrproject-rtos/zephyr/pull/20555) ?
(a link to TI website would work).
Ideally one day bindings description could be exported to doc.zephyr....

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by bwitherspoon** [comment]
> > Also are there any opinions on if this should be called BoosterPack connector instead of LaunchPad connector?
> 
> I think I have a preference for BoosterPack connector given it is referred to as "BoosterPack plug-in module connectors" on the [page for the TI LaunchPad](http://www.ti.com/tool/LAUNCHXL-CC1352R1). I also see some existing documentation (e.g. boards/arm/cc26x2r1_launchxl/doc/index.rst) where it is referred to as BoosterPack connector, so it is good to be consistent.

Agreed. I renamed this boosterpack.

File: dts/bindings/gpio/ti,launchpad-header.yaml (hunk 6 above)
Comment (bwitherspoon) [reply dts/bindings/gpio/ti,launchpad-header.yaml]: I added some documentation modeled after #20555.

File: dts/bindings/gpio/ti,boosterpack-header.yaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    BoosterPack plug-in modules are available in 20 and 40 pin variants. The
+    20 pin variant has two 10 x 1 pin headers and the 40 pin variant has two
+    10 x 2 pin headers. Both variants are compatible and stackable.
+
+    The pins of the 20 pin variant and the outer row of the 40 pin variant are
+    numbered 1 through 20. The inner rows of the 40 pin variant are numbered 21
+    through 40. The BoosterPack pinout is depicted below:
+
+    1  3.3V     21 5V                    40 GPIO 20 GND
+    2  Analog   22 GND                   39 GPIO 19 GPIO / SPI CS
```
Comment (vanti) [inline dts/bindings/gpio/ti,boosterpack-header.yaml]: Pin 18 should be "GPIO/SPI CS", not pin 19.

**Review by vanti — CHANGES_REQUESTED** [review]

File: dts/bindings/gpio/ti,boosterpack-header.yaml (hunk 7 above)
Comment (bwitherspoon) [reply dts/bindings/gpio/ti,boosterpack-header.yaml]: One off. Fixed.

**Review by vanti — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by bwitherspoon** [comment]
This change has already been merged indirectly.

**Pull request closed without merging**